src/utils/apps.ts

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {AppBinding} from '../types/apps';
+import {AppBindingLocations} from '../constants/apps';
+
+export function fillBindingsInformation(binding?: AppBinding) {
+    if (!binding) {
+        return;
+    }
+
+    binding.bindings?.forEach((b) => {
+        // Propagate id down if not defined
+        if (!b.app_id) {
+            b.app_id = binding.app_id;
+        }
+
+        // Compose location
+        b.location = binding.location + '/' + b.location;
+
+        // Propagate call down if not defined
+        if (!b.call) {
+            b.call = binding.call;
+        }
+
+        fillBindingsInformation(b);
+    });
+
+    // Trim branches without app_id
+    if (!binding.app_id) {
+        binding.bindings = binding.bindings?.filter((v) => v.app_id);
+    }
+
+    // Trim branches without calls
+    if (!binding.call) {
+        binding.bindings = binding.bindings?.filter((v) => v.call);
+    }
+
+    // Pull up app_id if needed
+    if (binding.bindings?.length && !binding.app_id) {
+        binding.app_id = binding.bindings[0].app_id;
+    }
+
+    // Pull up call if needed
+    if (binding.bindings?.length && !binding.call) {
+        binding.call = binding.bindings[0].call;
+    }
+}
+
+export function validateBindings(binding?: AppBinding) {
+    filterInvalidChannelHeaderBindings(binding);
+    filterInvalidCommands(binding);
+    filterInvalidPostMenuBindings(binding);
+    binding?.bindings?.forEach(fillBindingsInformation);
+}
+
+function filterInvalidCommands(binding?: AppBinding) {
+    if (!binding) {
+        return;
+    }
+
+    const isValidCommand = (b: AppBinding): boolean => {
+        return Boolean(b.label);
```

Comment:
We'll also need to document what makes a binding valid/invalid in yours/Michael's doc. For a command binding for example, is it just the presence of a label that makes it structurally valid (besides the "fixing" of other fields in `fillBindingsInformation `)?

Reply:
A valid leaf binding must have a call and an App ID.
A valid command must have a label, and be a valid binding.

I will document that.